src/python/ib1/openenergy/support/__init__.py

```diff
@@ -43,6 +46,236 @@ def httpclient_log(*args):
+                                    data=token_request)
+            if response.status_code == 200:
+                LOG.debug('Acquired authorization_code token')
+                j = response.json()
+                if 'refresh_token' in j:
+                    LOG.debug(f'Acquired refresh token')
+                    self._refresh_token = j['refresh_token']
+                return j
+            LOG.error(f'Unable to acquire authorization_code token : {response.json()}')
+            response.raise_for_status()
+
+        if self._refresh_token is None:
+            # No refresh token, use the code
+            return make_request(token_request={'client_id': client_id,
+                                               'scope': scopes,
+                                               'grant_type': 'authorization_code',
+                                               'code': self._code,
+                                               'code_verifier': self._verifier,
+                                               'redirect_uri': self.redirect_uri})
+        else:
+            # Use the refresh token, potentially updating it
+            return make_request(token_request={'client_id': client_id,
+                                               'scope': scopes,
+                                               'grant_type': 'refresh_token',
+                                               'refresh_token': self._refresh_token})
+
+
+class JWTBearerAuthMethod:
+    """
+    Additional auth logic passed to the FAPISession to enable jwt-bearer tokens
+    """
+
+    def __init__(self, jwt_bearer_email: str, signing_private_key: str, signing_key_password: str = None):
+        """
+        :param jwt_bearer_email:
+            This should be the email address of a user to impersonate. This will then
+            switch the client to jwt-bearer grant type. This will only work if the corresponding client has been
+            provisioned appropriately in the directory itself, otherwise this will fail. It should only ever be used
+            by our internal Open Energy clients needing to write to the directory, and can be entirely ignored by other
+            users.
+        :param signing_private_key:
+            Path to private key set as a signing key in the directory.
+        :param signing_key_password:
+            If specified, will be used as the password when loading the private signing key
+        """
+        self.jwt_bearer_email = jwt_bearer_email
+        with open(signing_private_key, 'rb') as key_file:
+            self.signing_private_key = serialization.load_pem_private_key(data=key_file.read(),
+                                                                          backend=default_backend(),
+                                                                          password=signing_key_password)
+
+    def get_token(self, session, client_id, scopes, openid_configuration):
```

Comment:
is it worth consistently adding assertions about the input types like in other methods?

Reply:
Probably yes! I have some other validation logic to add as well so will make a couple more commits to this branch (should update the PR automatically)